sklearn/preprocessing/_mdlp.pyx

```diff
@@ -0,0 +1,151 @@
+from __future__ import division
+from cpython.mem cimport PyMem_Malloc
+from cpython.mem cimport PyMem_Free
+import numpy as np
+cimport numpy as np
+from scipy.stats import entropy
+from libc.math cimport log, pow
+from libcpp.set cimport set as stdset
+from libcpp.vector cimport vector as stdvector
+from libc.math cimport INFINITY
+import cython
+
+cdef SIZE_t LEVEL_SIZE = sizeof(SIZE_t) * 3
+
+@cython.boundscheck(False)
+def MDLPDiscretize(col, y, bint shuffle, int min_depth):
+    """Performs the discretization process on X and y
+    """
+    # Shuffle array, and then reorder them
+    if shuffle:
+        shuffled_order = np.random.permutation(len(y))
+        col = col[shuffled_order]
+        y = y[shuffled_order]
+
+    order = np.argsort(col)
+    col = col[order]
+    y = y[order]
+
+    cdef stdset[FLOAT] cut_points = stdset[FLOAT]()
+
+    # Now we do a depth first search to create cut_points
+    cdef int num_samples = len(col)
+    cdef LEVEL init_level = <LEVEL> PyMem_Malloc(LEVEL_SIZE)
+    cdef stdvector[LEVEL] search_intervals = stdvector[LEVEL]()
+    set_level(init_level, 0, num_samples, 0)
+
+    search_intervals.push_back(init_level)
+    cdef LEVEL currlevel
+    while search_intervals.size() > 0:
+        currlevel = search_intervals.back()
+        search_intervals.pop_back()
+        start, end, depth = unwrap(currlevel)
+        PyMem_Free(currlevel)
```

Comment:
I'm only a beginner when it comes to Cython. Does Cython have a garbage collector, or will this cause a memory leak?